### DISCUSSION TITLE
Is there a reliable way to use the Prisma CLI to check if a new migration is needed?

### DISCUSSION THREAD

**Discussion by ChristopherZhong** [prisma/prisma#19418] [Q&A] [answered]
Our use case is as follows.
We have some developers who can change the Prisma schema. Sometimes, the changes do not require a new migration to be generated because they are simply doing a mapping (`map`). But sometimes they make mistakes like accidentally changing the index map value (e.g., `@@index([id], map: "abc123")`), which will require a new migration to be generated. So what I would like to do is to include some kind of checking in our CI/CD that will fail if a new migration needs to be generated.

Is there a reliable way to use the Prisma CLI to check if a new migration needs to be generated?

I am experimenting with a few different approaches. Here is one that I'm trying.

1. First, I'll count the number of directories in the `migrations` directory
2. Then I run `npx prisma migrate dev --create-only --name test`
3. I do another count
4. If the numbers do not match up, it fails

This approach feels hackish to me. All the other approaches that I can think of also feel hackish to me.

**Comment by nurul3101** [comment]
Hello @ChristopherZhong 👋 

Did you have a chance to look at the `migrate status` command: [Reference](https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-status)?

The prisma migrate status command looks up the migrations in /prisma/migrations/* folder and the entries in the _prisma_migrations table and compiles information about the state of the migrations in your database.

**Reply by ChristopherZhong** [reply]
I have. The `npx prisma migrate status` seems to report if there are any migrations that have not been applied. But I don't see how that would help in our case, since the developer forgot to generate the migration, it would say that the database schema is up to date.

**Reply by ChristopherZhong** [reply]
Here is my current approach.

```shell
#!/usr/bin/env bash
printf 'Applying existing migrations ...'
npx prisma migrate deploy
echo 'Done.'
printf 'Checking Prisma Schema ...'
string=$(npx prisma migrate dev --name test)
substring="Already in sync, no schema change or pending migration was found."
if [ "${string#*"$substring"}" != "$string" ]; then
    # no migration was generated
    echo "All good."
    exit 0
else
    # new migration was generated
    echo "A new migration was created!"
    echo "$string"
    exit 1
fi
```

**Reply by janpio** [reply]
You can probably use `migrate diff` to craft something a bit less heady and risky as `migrate dev`. It is the primitive that `migrate dev` uses under the hood anyway.

**Reply by Jolg42** [reply, marked as answer]
Maybe we should add that as an example in https://www.prisma.io/docs/reference/api-reference/command-reference#examples-14

It would look like this:

When checking the database directly:
```
npx prisma migrate diff \ 
--from-database-url "$DATABASE_URL"  \
--to-schema-datamodel ./schema.prisma \
--exit-code
```

or better using the migrations' directory 
```
npx prisma migrate diff \ 
--from-migrations ./migrations \
--to-schema-datamodel ./schema.prisma \
--exit-code
```
